File: src/main/scala/com/andbutso/largo/Note.scala
package com.andbutso.largo

import javax.sound.midi.{MidiSystem, MidiChannel}
import java.util.concurrent.{ConcurrentLinkedQueue, Executors}

import com.andbutso.largo

import scala.annotation.tailrec

trait PitchInterval {
  def intervals: Seq[SemiTone]
  def apply(note: Note): Seq[Note]
}

class AbsoluteInterval(val intervals: Seq[SemiTone]) extends PitchInterval {
  def apply(note: Note) = {
    intervals.foldLeft(Seq(note)) { case (notes, interval) =>
      notes :+ (note + interval)
    }
  }
}

object AbsoluteInterval {
  def apply(semiTones: SemiTone*) = new AbsoluteInterval(semiTones)
}

class StackedInterval(val intervals: Seq[SemiTone]) extends PitchInterval {
  def apply(note: Note) = {
    intervals.foldLeft(Seq(note)) { case (notes, interval) =>
      notes :+ (notes.last + interval)
    }
  }
}

object StackedInterval {
  def apply(semiTones: SemiTone*) = new StackedInterval(semiTones)
}

object PitchInterval {
  // Intervals of extended meantone temperament
  // Name                 Interval          Pitch (from C)   Roman No.
  val unison            = SemiTone(0)       // C             I
  val diminishedSecond  = unison            // D♭♭
  val chromaticSemiTone = SemiTone(1)       // C♯
  val minorSecond       = chromaticSemiTone // D♭
  val halfTone          = minorSecond
  val wholeTone         = SemiTone(2)       // D             II
  val diminishedThird   = wholeTone         // E♭♭
  val augmentedSecond   = SemiTone(3)       // D♯
  val minorThird        = augmentedSecond   // E♭
  val majorThird        = SemiTone(4)       // E             III
  val diminishedFourth  = majorThird        // F♭
  val augmentedThird    = SemiTone(5)       // E♯
  val perfectFourth     = augmentedThird    // F             IV
  val augmentedFourth   = SemiTone(6)       // F♯
  val diminishedFifth   = augmentedFourth   // G♭
  val perfectFifth      = SemiTone(7)       // G             V
  val diminishedSixth   = perfectFifth      // A♭♭
  val augmentedFifth    = SemiTone(8)       // G♯
  val minorSixth        = augmentedFifth    // A♭
  val majorSixth        = SemiTone(9)       // A             VI
  val diminishedSeventh = majorSixth        // B♭♭
  val augmentedSixth    = SemiTone(10)      // A♯
  val minorSeventh      = augmentedSixth    // B♭
  val majorSeventh      = SemiTone(11)      // B             VII
  val diminishedOctave  = majorSeventh      // C♭
  val augmentedSeventh  = SemiTone(12)      // B♯
  val octave            = augmentedSeventh  // C             VIII
}

object Accidental extends Enumeration {
  val Natural, Sharp, Flat, DoubleSharp, DoubleFlat = Value
}

// TODO DiatonicScale

object Scale {
  import PitchInterval._

  val major = StackedInterval(
    wholeTone, // Supertonic
    wholeTone, // Mediant
    halfTone,  // Subdominant
    wholeTone, // Dominant
    wholeTone, // Submediant
    wholeTone, // Leading tone
    halfTone   // Tonic / Octave
  )
  // TODO Natural minor scale vs harmonic minor scale (https://en.wikipedia.org/wiki/Minor_scale)
  val minor = StackedInterval(
    wholeTone,
    halfTone,
    wholeTone,
    wholeTone,
    halfTone,
    wholeTone,
    wholeTone
  )

  object Degree extends Enumeration {
    val Tonic       = Value("I")
    val Supertonic  = Value("ii")
    val Mediant     = Value("iii")
    val Subdominant = Value("IV")
    val Dominant    = Value("V")
    val Submediant  = Value("vi")
    val Subtonic    = Value("vii")
  }
}

abstract class Scale(tonicKeyNote: Note, intervalsFromTonicKeyNote: PitchInterval) {
  val degrees = tonicKeyNote ++ intervalsFromTonicKeyNote
  def supertonic  = degrees(1)
  def mediant     = degrees(2)
  def subdominant = degrees(3)
  def dominant    = degrees(4)
  def submediant  = degrees(5)
  def leadingTone = degrees(6)
  def octave      = degrees(7)
}

case class MajorScale(tonicKeyNote: Note) extends Scale(tonicKeyNote, Scale.major)
case class MinorScale(tonicKeyNote: Note) extends Scale(tonicKeyNote, Scale.minor)

import Accidental._

// TODO These should be chords rather than pitches
// (in other words, the perfect fifth interval should be calculated from the tonic of each chord)
case class Key(pitch: Pitch) {
  import PitchInterval.perfectFifth

  val IV  = pitch - perfectFifth
  val I   = pitch
  val V   = pitch + perfectFifth
  val ii  = V     + perfectFifth
  val vi  = ii    + perfectFifth
  val iii = vi    + perfectFifth
  val vii = iii   + perfectFifth

  val major      = Seq(IV, I, V)
  val minor      = Seq(ii, vi, iii)
  val diminished = Seq(vii)

  val chords = major ++ minor ++ diminished
}

// https://en.wikipedia.org/wiki/Seventh_chord
//
// Tertian:
// Cmaj⁷   — Major seventh (major/major)
// Cmin⁷   — Minor seventh (minor/minor)
// C⁷      — Dominant seventh (major/minor)
// C°⁷     — Diminished seventh
// Cm⁷     — Half-diminished seventh (diminished/minor)
// Cm maj⁷ - Minor major seventh
// Cmaj⁷⁽⁵⁾  – Augmented major seventh
//
// Non-tertian:
// Caug⁷   — Augmented seventh (augmented/minor)
// CmM7♭5 — Diminished major seventh
// C7♭5   — Dominant seventh flat five
case class Seventh(triad: Triad, seventh: Note)

// TODO Given a set of notes be able to detect based on the interval between them
// if they are any of the defined Triads or Sevenths or scales

object Seventh {
  import PitchInterval._

  val dominant = AbsoluteInterval(
    majorThird,
    perfectFifth,
    minorSeventh
  )

  val major = AbsoluteInterval(
    majorThird,
    perfectFifth,
    majorSeventh
  )

  val minorMajor = AbsoluteInterval(
    minorThird,
    perfectFifth,
    majorSeventh
  )

  val minor = AbsoluteInterval(
    minorThird,
    perfectFifth,
    minorSeventh
  )

  val augmentedMajor = AbsoluteInterval(
    majorThird,
    augmentedFifth,
    majorSeventh
  )

  val augmented = AbsoluteInterval(
    majorThird,
    augmentedFifth,
    minorSeventh
  )

  val halfDiminished = AbsoluteInterval(
    minorThird,
    diminishedFifth,
    minorSeventh
  )

  val diminished = AbsoluteInterval(
    minorThird,
    diminishedFifth,
    diminishedSeventh
  )

  val dominantFlatFive = AbsoluteInterval(
    majorThird,
    diminishedFifth,
    minorSeventh
  )
}

// TODO Altered chords (https://en.wikipedia.org/wiki/Chord_(music)#Altered_chords)
// TODO Suspended chords (https://en.wikipedia.org/wiki/Chord_(music)#Suspended_chords)

object Chord {
  object Quality extends Enumeration {
    val Major          = Value("maj") // Δ, M
    val Minor          = Value("min") // -
    val Diminished     = Value("dim") // o, °
    val HalfDiminished = Value("ø")   // ø, Ø
    val Dominant       = Value("dom") // 7
    val Augmented      = Value("aug") // +
  }

  object IntervalNumber extends Enumeration {
    val Third         = Value
    val Triad         = Third
    val Fifth         = Value("⁵")
    val Indeterminate = Fifth
    val Neutral       = Fifth
    val Sixth         = Value("⁶")
    val Seventh       = Value("⁷")
    val Ninth         = Value("⁹")
    val Eleventh      = Value("¹¹")
    val Thirteenth    = Value("¹³")
  }

  object Category extends Enumeration {
    // Secundal chords can be decomposed into a series of (major or minor) seconds.
    // For example, the chord C-D-E♭ is a series of seconds, containing a major second (C-D) and a minor second (D-E♭)
    val Secundal = Value // 2nd's: major 2nd, minor 2nd
    // Tertian chords can be decomposed into a series of (major or minor) thirds.
    // For example, the C major triad (C-E-G) is defined by a sequence of two intervals, the first (C-E) being
    // a major third and the second (E-G) being a minor third. Most common chords are tertian.
    val Tertian  = Value // 3rd's: major 3rd, minor 3rd
    // Quartal chords can be decomposed into a series of (perfect or augmented) fourths.
    // Quartal harmony normally works with a combination of perfect and augmented fourths.
    // Diminished fourths are enharmonically equivalent to major thirds, so they are uncommon.
    // For example, the chord C-F-B is a series of fourths, containing a perfect fourth (C-F) and
    // an augmented fourth/tritone (F-B).
    val Quartal  = Value // 4th's: perfect 4th, augmented 4th
    val Quintal  = Value // 5th's
  }
}
// "A chord is a combination of three or more tones sounded simultaneously"
case class Chord(
  root: Note,
  quality: Chord.Quality.Value,
  number: Chord.IntervalNumber.Value
) {
  // TODO Implement inversion
  def inversion(number: Int) = ???
}

object Triad {
  import PitchInterval._

  val major = StackedInterval(
    majorThird, // 4 semitones from root
    minorThird  // perfect fifth, 7 semitones from root
  )

  val minor = StackedInterval(
    minorThird, // 3 semitones from root
    majorThird  // perfect fifth, 7 semitones from root
  )

  val diminished = StackedInterval(
    minorThird, // 3 semitones from root
    minorThird  // diminished fifth, 6 semitones from root
  )

  val augmented = StackedInterval(
    majorThird, // 4 semitones from root
    majorThird  // augmented fifth, 8 semitones from root
  )
}

case class Triad(notes: IndexedSeq[Note]) {
  def root  = notes(0)
  def third = notes(1)
  def fifth = notes(2)

  // TODO Implement
  def inversion(number: Int) = ???
}

case class Pitch(
  letter: Char,
  accidental: Accidental.Value = Natural
) {
  def ♯ = {
    copy(accidental = Sharp)
  }

  def ♭ = {
    copy(accidental = Flat)
  }

  def ♮ = {
    copy(accidental = Natural)
  }

  def apply(octave: Octave) = {
    Note(this, octave)
  }

  def +(number: Int): Pitch = {
    this + SemiTone(number)
  }

  def +(semiTone: SemiTone): Pitch = {
    PitchClass.moveUpFrom(this, semiTone)
  }

  def -(number: Int): Pitch = {
    this - SemiTone(number)
  }

  def -(semiTone: SemiTone): Pitch = {
    PitchClass.moveDownFrom(this, semiTone)
  }

  def interval(pitch: Pitch) = {
    SemiTone(PitchClass.indexOf(pitch) - PitchClass.indexOf(this))
  }

  override def toString = {
    val accidentalString = accidental match {
      case Sharp       => "♯"
      case Flat        => "♭"
      case Natural     => ""
      case DoubleFlat  => "♭♭" // N.B. Double flat unicode doesn't work in IntelliJ
      case DoubleSharp => "♯♯" // N.B. Double sharp unicode doesn't work either
    }

    s"$letter$accidentalString"
  }
}

object Pitch {
  val C = Pitch('C')
  val D = Pitch('D')
  val E = Pitch('E')
  val F = Pitch('F')
  val G = Pitch('G')
  val A = Pitch('A')
  val B = Pitch('B')

  object Frequency {
    val C0 = 16.35D
    val A4 = 440.0D
    val ConcertPitch = A4
    val B8 = 7902.13D
  }
}

import Pitch._

object PitchClass extends PitchClass(
  IndexedSeq(
    C,
    C♯,
    D,
    D♯,
    E,
    F,
    F♯,
    G,
    G♯,
    A,
    A♯,
    B
  )
)

case class PitchClass(pitches: IndexedSeq[Pitch]) {
  val indexes = pitches.zipWithIndex.toMap

  def moveUpFrom(pitch: Pitch, semiTone: SemiTone) = {
    val index         = indexOf(pitch)
    val indexToMoveTo = (index + semiTone.interval) % 12
    pitches(indexToMoveTo)
  }

  def moveDownFrom(pitch: Pitch, semiTone: SemiTone) = {
    val index         = indexOf(pitch)
    val indexToMoveTo = (index - semiTone.interval) % 12
    val destination = if (indexToMoveTo < 0) pitches.size + indexToMoveTo else indexToMoveTo
    pitches(destination)
  }

  def moveFrom(pitch: Pitch, semiTone: SemiTone) = {
    semiTone.interval match {
      case 0 =>
        pitch
      case interval if interval > 0 =>
        moveUpFrom(pitch, semiTone)
      case interval =>
        moveDownFrom(pitch, SemiTone(-semiTone.interval))
    }
  }

  @tailrec
  final def indexOf(pitch: Pitch): Int = {
    pitch.accidental match {
      case Flat =>
        val natural = pitch.♮
        val index   = indexes(natural)
        if (index != 0) {
          indexes(natural) - 1
        } else {
          indexOf(B)
        }
      case _ =>
        indexes(pitch)
    }
  }
}

object SemiTone {
  val cents = 1200
  val perOctave = 12

  // e.g.
  //  val freqA = 130.81 (C3)
  //  val freqB = 261.63 (C4)
  //  intervalBetweenFrequencies(freqA, freqB)
  //  => SemiTone(12)
  def intervalBetweenFrequencies(freqA: Hertz, freqB: Hertz) = {
    SemiTone(Math.round(cents * Frequency.logarithmicRatio(freqA, freqB) / 100).toInt)
  }

  implicit def fromSemiToneToInt(semiTone: SemiTone): Int = semiTone.interval
  implicit def fromIntToSemiTone(int: Int): SemiTone      = SemiTone(int)
}

case class SemiTone(interval: Int)

object Note {
  val all = Octave.Min.to(Octave.Max).flatMap { octave =>
    PitchClass.pitches.map { pitch =>
      Note (pitch, octave)
    }
  }

  val ConcertPitch = Note(A, 4)

  def fromFrequency(frequency: Hertz) = {

  }
}

case class Note(pitch: Pitch, octave: Octave) extends Ordered[Note] {
  def toFrequency = {
    val intervalFromConcertPitch = Note.ConcertPitch.interval(this)
    Pitch.Frequency.ConcertPitch * Math.pow(Octave.TwelfthRootOf2, intervalFromConcertPitch.interval)
  }

  def index = {
    PitchClass.indexOf(pitch) + (octave.value * SemiTone.perOctave)
  }

  override def equals(obj: Any) = {
    obj match {
      case o: Note => o.index.equals(index)
      case _ => false
    }
  }

  override def hashCode = index.hashCode()

  def interval(otherNote: Note) = {
    val octaveInterval = octave.interval(otherNote.octave) * SemiTone.perOctave
    SemiTone(pitch.interval(otherNote.pitch).interval + octaveInterval)
  }

  def compare(otherNote: Note) = {
    0.compare(interval(otherNote).interval)
  }

  def +(interval: SemiTone) = {
    Note.all(index + interval.interval)
  }

  def ++(intervals: PitchInterval) = {
    intervals(this)
  }

  def majorScale = {
    this ++ Scale.major
  }

  def minorScale = {
    this ++ Scale.minor
  }

  override def toString = {
    s"$pitch${octave.value}"
  }
}

object Octave {
  val TwelfthRootOf2 = Math.pow(2, 1/12.0)
  val Min = 0
  val Max = 8

  implicit def fromOctaveToInt(octave: Octave): Int = octave.value
  implicit def fromIntToOctave(int: Int): Octave    = Octave(int)
}

// Octave is two notes that have a frequency ration of 2:1
// Octave spans 12 semitones or 1200 cents
case class Octave(value: Int) {
  require(value >= Octave.Min && value <= Octave.Max, s"$value out of bounds")

  def interval(octave: Octave) = {
    octave.value - value
  }

  def apply(pitch: Pitch) = {
    Note(pitch, this)
  }
}

// TODO
// KeySignature
// TimeSignature

File: note.go
package largo

import (
	"fmt"

	"github.com/marcel/largo/chord"
)

const (
	ConcertPitch       = 440.0
	SemiToneCents      = 1200
	SemiTonesPerOctave = 12
)

type SemiTone uint8

const (
	// Intervals of extended meantone temperament
	// Name                 Interval          Pitch (from C)   Roman No.
	Unison            = SemiTone(0)       // C             I
	DiminishedSecond  = Unison            // D♭♭
	ChromaticSemiTone = SemiTone(1)       // C♯
	MinorSecond       = ChromaticSemiTone // D♭
	HalfTone          = MinorSecond
	WholeTone         = SemiTone(2)      // D             II
	DiminishedThird   = WholeTone        // E♭♭
	AugmentedSecond   = SemiTone(3)      // D♯
	MinorThird        = AugmentedSecond  // E♭
	MajorThird        = SemiTone(4)      // E             III
	DiminishedFourth  = MajorThird       // F♭
	AugmentedThird    = SemiTone(5)      // E♯
	PerfectFourth     = AugmentedThird   // F             IV
	AugmentedFourth   = SemiTone(6)      // F♯
	DiminishedFifth   = AugmentedFourth  // G♭
	PerfectFifth      = SemiTone(7)      // G             V
	DiminishedSixth   = PerfectFifth     // A♭♭
	AugmentedFifth    = SemiTone(8)      // G♯
	MinorSixth        = AugmentedFifth   // A♭
	MajorSixth        = SemiTone(9)      // A             VI
	DiminishedSeventh = MajorSixth       // B♭♭
	AugmentedSixth    = SemiTone(10)     // A♯
	MinorSeventh      = AugmentedSixth   // B♭
	MajorSeventh      = SemiTone(11)     // B             VII
	DiminishedOctave  = MajorSeventh     // C♭
	AugmentedSeventh  = SemiTone(12)     // B♯
	OctaveTone        = AugmentedSeventh // C             VIII
)

type PitchInterval interface {
	FromNote(Note) Notes
}

type StackedInterval []SemiTone

func (s StackedInterval) FromNote(note Note) Notes {
	// intervals.foldLeft(Seq(note)) { case (notes, interval) =>
	//   notes :+ (notes.last + interval)
	// }
	panic("not implemented")
}

type AbsoluteInterval []SemiTone

func (a AbsoluteInterval) FromNote(note Note) Notes {
	// intervals.foldLeft(Seq(note)) { case (notes, interval) =>
	//   notes :+ (note + interval)
	// }
	panic("not implemented")
}

// TODO implement sort.Interface
type Note struct {
	Pitch  Pitch
	Octave Octave
}

func NewNote(pitch Pitch, octave Octave) Note {
	return Note{
		Pitch:  pitch,
		Octave: octave,
	}
}

func (n Note) Frequency() {
	// val intervalFromConcertPitch = Note.ConcertPitch.interval(this)
	// Pitch.Frequency.ConcertPitch * Math.pow(Octave.TwelfthRootOf2, intervalFromConcertPitch.interval)
	panic("not implemented")
}

func (n Note) Index() int {
	// CircularPitchClass.indexOf(pitch) + (octave.value * SemiTone.perOctave)
	panic("not implemented")
}

// override func (n Note) equals(obj: Any) = {
//   obj match {
//     case o: Note => o.index.equals(index)
//     case _ => false
//   }
// }

func (n Note) Interval(otherNote Note) SemiTone {
	// val octaveInterval = octave.interval(otherNote.octave) * SemiTone.perOctave
	// SemiTone(pitch.interval(otherNote.pitch).interval + octaveInterval)
	panic("not implemented")
}

// func (n Note) compare(otherNote: Note) = {
//   0.compare(interval(otherNote).interval)
// }

func (n Note) Raise(interval SemiTone) Note {
	return PitchSpace[n.Index()+int(interval)]
}

// TODO this was called ++ in Scala, so I'm  not sure if the name I chose is
// correct
func (n Note) RaiseAccordingTo(intervals PitchInterval) Notes {
	return intervals.FromNote(n)
}

func (n Note) MajorScale() Notes {
	return n.RaiseAccordingTo(MajorScale)
}

func (n Note) MinorScale() Notes {
	return n.RaiseAccordingTo(MinorScale)
}

func (n Note) String() string {
	return fmt.Sprintf("%s%d", n.Pitch.String(), n.Octave)
}

type Notes []Note

func (s SemiTone) IntervalBetweenFrequencies() {
	panic("not implemented yet")
}

type Accidental uint8

const (
	Natural Accidental = iota + 1
	Sharp
	Flat
	DoubleSharp
	DoubleFlat
)

type Degree string

const (
	Tonic       = Degree("I")
	Supertonic  = Degree("ii")
	Mediant     = Degree("iii")
	Subdominant = Degree("IV")
	Dominant    = Degree("V")
	Submediant  = Degree("vi")
	Subtonic    = Degree("vii")
)

var (
	MajorScale = AbsoluteInterval{
		WholeTone, // Supertonic
		WholeTone, // Mediant
		HalfTone,  // Subdominant
		WholeTone, // Dominant
		WholeTone, // Submediant
		WholeTone, // Leading tone
		HalfTone,  // Tonic / Octave
	}

	// TODO Natural minor scale vs harmonic minor scale (https://en.wikipedia.org/wiki/Minor_scale)
	MinorScale = AbsoluteInterval{
		WholeTone,
		HalfTone,
		WholeTone,
		WholeTone,
		HalfTone,
		WholeTone,
		WholeTone,
	}
)

type Scale struct {
	TonicKeyNote              Note
	IntervalsFromTonicKeyNote PitchInterval
}

func (s Scale) Degrees() Notes    { return s.TonicKeyNote.RaiseAccordingTo(s.IntervalsFromTonicKeyNote) }
func (s Scale) Supertonic() Note  { return s.Degrees()[1] }
func (s Scale) Mediant() Note     { return s.Degrees()[2] }
func (s Scale) Subdominant() Note { return s.Degrees()[3] }
func (s Scale) Dominant() Note    { return s.Degrees()[4] }
func (s Scale) Submediant() Note  { return s.Degrees()[5] }
func (s Scale) LeadingTone() Note { return s.Degrees()[6] }
func (s Scale) Octave() Note      { return s.Degrees()[7] }

type Letter rune

func (l Letter) Pitch(optionalAccidental ...Accidental) Pitch {
	accidental := Natural

	if optionalAccidental != nil {
		accidental = optionalAccidental[0]
	}

	return Pitch{
		Letter:     l,
		Accidental: accidental,
	}
}

type Pitch struct {
	Letter     Letter
	Accidental Accidental
}

func (p Pitch) AtOctave(octave Octave) Note {
	return NewNote(p, octave)
}

func (p Pitch) Sharp() Pitch {
	return p.Letter.Pitch(Sharp)
}

func (p Pitch) Flat() Pitch {
	return p.Letter.Pitch(Flat)
}

func (p Pitch) Natural() Pitch {
	return p.Letter.Pitch(Natural)
}

func (p Pitch) Note(octave Octave) Note {
	// return NewNote(p, octave)
	panic("not implemented")
}

func (p Pitch) RaiseBy(number int) Pitch {
	// this + SemiTone(number)
	panic("not implemented")
}

func (p Pitch) MoveUpFrom(semiTone SemiTone) Pitch {
	// CircularPitchClass.moveUpFrom(this, semiTone)
	panic("not implemented")
}

func (p Pitch) LowerBy(number int) Pitch {
	// this - SemiTone(number)
	panic("not implemented")
}

func (p Pitch) MoveDownFrom(semiTone SemiTone) Pitch {
	// CircularPitchClass.moveDownFrom(this, semiTone)
	panic("not implemented")
}

func (p Pitch) Interval(pitch Pitch) {
	// SemiTone(CircularPitchClass.indexOf(pitch) - CircularPitchClass.indexOf(this))
	panic("not implemented, figure out return type")
}

func (p Pitch) String() string {
	display := func(s string) string {
		return string(p.Letter) + string(p.Accidental)
	}
	switch p.Accidental {
	case Sharp:
		return display("♯")
	case Flat:
		return display("♭")
	case Natural:
		return display("")
	case DoubleFlat:
		return display("♭♭") // N.B. Double flat unicode work?
	case DoubleSharp:
		return display("♯♯") // N.B. Double sharp unicode work?
	}

	return "<unknown pitch>"
}

// TODO replace with whatever resolves the TODO for Key
type Pitches []Pitch

// TODO These should be chords rather than pitches
// (in other words, the perfect fifth interval should be calculated from the tonic of each chord)
type Key Pitch

func (k Key) Pitch() Pitch { return Pitch(k) }
func (k Key) IV() Pitch    { return k.Pitch().MoveDownFrom(PerfectFifth) }
func (k Key) I() Pitch     { return Pitch(k) }
func (k Key) V() Pitch     { return k.Pitch().MoveUpFrom(PerfectFifth) }

// N.B. II - VII should be lower case in accordance with proper musical
// notation but we want these methods to be exported to we have capitalized
// them
func (k Key) II() Pitch  { return k.V().MoveUpFrom(PerfectFifth) }
func (k Key) VI() Pitch  { return k.II().MoveUpFrom(PerfectFifth) }
func (k Key) III() Pitch { return k.VI().MoveUpFrom(PerfectFifth) }
func (k Key) VII() Pitch { return k.III().MoveUpFrom(PerfectFifth) }

func (k Key) Major() Pitches      { return Pitches{k.IV(), k.I(), k.V()} }
func (k Key) Minor() Pitches      { return Pitches{k.II(), k.VI(), k.III()} }
func (k Key) Diminished() Pitches { return Pitches{k.VII()} }

func (k Key) Chords() Pitches { return append(k.Major(), append(k.Minor(), k.Diminished()...)...) }

var PitchSpace = func() []Note {
	// Octave.Min.to(Octave.Max).flatMap { octave =>
	//   CircularPitchClass.pitches.map { pitch =>
	//     Note(pitch, octave)
	//   }
	// }
	panic("not implemented")
}()

var (
	C = Letter('C').Pitch()
	D = Letter('D').Pitch()
	E = Letter('E').Pitch()
	F = Letter('F').Pitch()
	G = Letter('G').Pitch()
	A = Letter('A').Pitch()
	B = Letter('B').Pitch()
)

// https://en.wikipedia.org/wiki/Seventh_chord
//
// Tertian:
// Cmaj⁷   — Major seventh (major/major)
// Cmin⁷   — Minor seventh (minor/minor)
// C⁷      — Dominant seventh (major/minor)
// C°⁷     — Diminished seventh
// Cm⁷     — Half-diminished seventh (diminished/minor)
// Cm maj⁷ - Minor major seventh
// Cmaj⁷⁽⁵⁾  – Augmented major seventh
//
// Non-tertian:
// Caug⁷   — Augmented seventh (augmented/minor)
// CmM7♭5 — Diminished major seventh
// C7♭5   — Dominant seventh flat five
type Seventh struct {
	Triad Triad
	Note  Note
}

var (
	DominantSeventh = AbsoluteInterval{
		MajorThird,
		PerfectFifth,
		MinorSeventh,
	}

	MajorSeventhSeventh = AbsoluteInterval{
		MajorThird,
		PerfectFifth,
		MajorSeventh,
	}

	MinorMajorSeventh = AbsoluteInterval{
		MinorThird,
		PerfectFifth,
		MajorSeventh,
	}

	MinorSeventhSeventh = AbsoluteInterval{
		MinorThird,
		PerfectFifth,
		MinorSeventh,
	}

	AugmentedMajorSeventh = AbsoluteInterval{
		MajorThird,
		AugmentedFifth,
		MajorSeventh,
	}

	AugmentedSeventhSeventh = AbsoluteInterval{
		MajorThird,
		AugmentedFifth,
		MinorSeventh,
	}

	HalfDiminishedSeventh = AbsoluteInterval{
		MinorThird,
		DiminishedFifth,
		MinorSeventh,
	}

	DiminishedSeventhSeventh = AbsoluteInterval{
		MinorThird,
		DiminishedFifth,
		DiminishedSeventh,
	}

	DominantFlatFiveSeventh = AbsoluteInterval{
		MajorThird,
		DiminishedFifth,
		MinorSeventh,
	}
)

type Triad [3]Note

func (t Triad) Root() Note  { return t[0] }
func (t Triad) Third() Note { return t[1] }
func (t Triad) Fifth() Note { return t[2] }

func (t Triad) Inversion(number int) {
	panic("not implemented")
}

var (
	MajorTriad = StackedInterval{
		MajorThird, // 4 semitones from root
		MinorThird, // perfect fifth, 7 semitones from root
	}

	MinorTriad = StackedInterval{
		MinorThird, // 3 semitones from root
		MajorThird, // perfect fifth, 7 semitones from root
	}

	DiminishedTriad = StackedInterval{
		MinorThird, // 3 semitones from root
		MinorThird, // diminished fifth, 6 semitones from root
	}

	AugmentedTriad = StackedInterval{
		MajorThird, // 4 semitones from root
		MajorThird, // augmented fifth, 8 semitones from root
	}
)

const (
	TwelfthRootOf2 = 1.0594630943592953 // 2^(1/12)
	LowestOctave   = 0
	HighestOctave  = 8
)

// Octave is two notes that have a frequency ratio of 2:1. An octave spans 12
// semitones or 1200 cents.
type Octave uint8

func (o Octave) Interval(otherOctave Octave) Octave {
	return otherOctave - o
}

func (o Octave) Note(pitch Pitch) Note {
	return NewNote(pitch, o)
}

// TODO Altered chords (https://en.wikipedia.org/wiki/Chord_(music)#Altered_chords)
// TODO Suspended chords (https://en.wikipedia.org/wiki/Chord_(music)#Suspended_chords)
// a Chord is a combination of three or more tones sounded simultaneously
type Chord struct {
	Root    Note
	Quality chord.Quality
	Number  chord.IntervalNumber
}

var CircularPitchClass = Pitches{
	C,
	C.Sharp(),
	D,
	D.Sharp(),
	E,
	F,
	F.Sharp(),
	G,
	G.Sharp(),
	A,
	A.Sharp(),
	B,
}

func (p Pitches) Indexes() map[Pitch]int {
	// p.zipWithIndex.toMap
	panic("not implemented")
}

func (p Pitches) MoveUpFrom(pitch Pitch, semiTone SemiTone) Pitch {
	index := p.indexOf(pitch)
	indexToMoveTo := (index + int(semiTone)) % 12
	return p[indexToMoveTo]
}

func (p Pitches) MoveDownFrom(pitch Pitch, semiTone SemiTone) Pitch {
	index := p.indexOf(pitch)
	indexToMoveTo := (index - int(semiTone)) % 12
	destination := indexToMoveTo
	if indexToMoveTo < 0 {
		destination = len(p) + indexToMoveTo
	}
	return p[destination]
}

func (p Pitches) MoveFrom(pitch Pitch, semiTone SemiTone) Pitch {
	if semiTone == 0 {
		return pitch
	}

	if semiTone > 0 {
		return p.MoveUpFrom(pitch, semiTone)
	}

	return p.MoveDownFrom(pitch, SemiTone(-semiTone))
}

func (p Pitches) indexOf(pitch Pitch) int {
	// switch pitch.Accidental {
	// case Flat :
	//     val natural = pitch.♮
	//     val index   = indexes(natural)
	//     if (index != 0) {
	//       indexes(natural) - 1
	//     } else {
	//       indexOf(B)
	//     }
	// default:
	//     return indexes(pitch)
	// }
	panic("not implmeented")
}
